### DISCUSSION TITLE
namedtuple vs dataclasses

### DISCUSSION THREAD

**Discussion by grmnz** [aliev/aioauth#57] [Ideas]
Hi!
First, congrats on the good library you have developed! Thanks)

I'd like to hear your opinion on the usage of `dataclass`'s instead of `NamedTuple`s on the project.
One of the reasons that I want to set a user to the request instance after the authorization process, but currently it is not possible.
There are also some difficulties with inheritance from `NamedTuple`

I suggest that it would be more convenient to write something like

```python
from aioauth.requests import Request as _Request

@dataclass(slots=True)
class Request(_Request):
   extra: Optional[dict] = None


request = Request(**payload)
request.user = User()

```

@aliev, what do you think about using `dataclass`'s?

P.S. Some notes from [SO](https://stackoverflow.com/questions/51671699/data-classes-vs-typing-namedtuple-primary-use-cases)

> When your data structure needs to/can be immutable, hashable, iterable, unpackable, comparable then you can use NamedTuple. If you need something more complicated, for example, a possibility of inheritance for your data structure then use Dataclass.

**Comment by aliev** [comment]
Hi @grmnz thank you!

I'm not against of the data classes if it will increase the extensibility of the project. I will test this idea soon.

Regarding of setting up the user to the request object you can take a look an example with FastAPI: https://github.com/aliev/aioauth-fastapi/blob/master/aioauth_fastapi/utils.py.

**Comment by aliev** [comment]
@grmnz [v1.4.0](https://github.com/aliev/aioauth/releases/tag/v1.4.0) release on dataclasses now! thanks for your contribution!
